### PR TITLE
Add trigger handler for storage events

### PR DESCRIPTION
**Pull request by bombguy**
## Problem

add ability to add trigger for storage events
**Issue number, if available:**

## Changes

**Corresponding docs PR, if applicable:**

## Checklist

- [ ] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [ ] If this PR requires a docs update, I have linked to that docs PR above.
- [ ] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/backend-storage/src/construct.ts [hunk 1]
```diff
@@ -61,6 +71,17 @@ export class AmplifyStorage
...
   }
 
+  /**
+   * Attach a Lambda function trigger handler to the S3 events
+   * @param events - list of S3 events that will trigger the handler
+   * @param handler - The function that will handle the event
+   */
+  addTrigger = (events: EventType[], handler: IFunction): void => {
+    handler.addEventSource(
+      new S3EventSource(this.resources.bucket as Bucket, { events })
```
[the remark this answers is not in the document]
Comment (bombguy) [reply packages/backend-storage/src/construct.ts:81]: We would need to bump the version of CDK to get this change: https://github.com/aws/aws-cdk/pull/28943

This change is on CDK version [2.127.0](https://github.com/aws/aws-cdk/blob/main/CHANGELOG.v2.md#21270-2024-02-09)

Should we consider bumping this in this PR?

[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/backend-storage/src/construct.ts:81]: yeah we can bump it

[the remark this answers is not in the document]
Comment (bombguy) [reply packages/backend-storage/src/construct.ts:81]: should we bump it in all the vertical packages or just in backend-storage?
Screenshot 2024-02-14 at 10 33 05 AM

File: .changeset/khaki-panthers-grin.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+'@aws-amplify/backend-storage': minor
+'@aws-amplify/integration-tests': patch
```
Comment (edwardfoyle) [inline .changeset/khaki-panthers-grin.md:3]: we need to add all the other packages that have changes to this list

**Comment by edwardfoyle** [comment]
LGTM, but the dependency check is complaining that some version of `aws-cdk` does not match the version of `aws-cdk-lib`. Also, merging from main may fix the changeset check

File: .changeset/five-fireants-shout.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+'@aws-amplify/backend-platform-test-stubs': minor
+'@aws-amplify/backend-output-storage': minor
+'@aws-amplify/integration-tests': minor
+'@aws-amplify/backend-function': minor
+'@aws-amplify/backend-storage': minor
+'@aws-amplify/auth-construct-alpha': minor
+'@aws-amplify/backend-auth': minor
+'@aws-amplify/backend-data': minor
+'@aws-amplify/plugin-types': minor
+'@aws-amplify/backend': minor
```
Comment (edwardfoyle) [inline .changeset/five-fireants-shout.md:11]:
```suggestion
'@aws-amplify/backend-platform-test-stubs': patch
'@aws-amplify/backend-output-storage': patch
'@aws-amplify/integration-tests': patch
'@aws-amplify/backend-function': patch
'@aws-amplify/backend-storage': patch
'@aws-amplify/auth-construct-alpha': patch
'@aws-amplify/backend-auth': patch
'@aws-amplify/backend-data': patch
'@aws-amplify/plugin-types': patch
'@aws-amplify/backend': patch
```

**Review by 0618 — APPROVED** [review]

**Review by edwardfoyle — APPROVED** [review]

**Pull request merged**